pupil_src/shared_modules/video_capture/utils.py

```diff
@@ -155,3 +165,144 @@ def check_slice(self, frame_gray):
+
+    @property
+    def base(self) -> str:
+        return os.path.split(self.path)[0]
+
+    @property
+    def timestamps(self) -> np.ndarray:
+        if self.ts is None:
+            self.load_ts()
+        return self.ts
+
+
+class VideoSet:
+    def __init__(self, rec: str, name: str):
+        self.rec = rec
+        self.name = name
+        self.video_exts = ("mp4", "mjpeg", "h264")
+        self._videos = None
+
+    @property
+    def videos(self) -> Sequence[Video]:
+        if self._videos is None:
+            self._videos = sorted(self.fetch_videos(), key=lambda v: v.path)
+        return self._videos
+
+    @property
+    def containers(self) -> Sequence[Video]:
+        return [av.open(video.path) for video in self.videos]
+
+    @property
+    def lookup_loc(self) -> str:
+        return os.path.join(self.rec, f"{self.name}_lookup.npy")
+
+    def fetch_videos(self) -> Iterator[Video]:
+        yield from (
+            Video(loc)
+            for ext in self.video_exts
+            for loc in glob.iglob(
+                os.path.join(self.rec, f"{self.name}*.{ext}"))
+        )
+
+    def build_lookup(self):
+        """
+        The lookup table is a np.recarray containing entries
+        for each (virtual and real) frame.
+
+        Each entry consists of 3 values:
+            - container_idx: Corresponding self.videos index
+            - container_frame_idx: Frame index within the container
+            - timestamp: Recorded or virtual Pupil timestamp
+
+        container_idx entries of value -1 indicate a virtual frame.
+
+        The lookup table can be easiliy filtered for real frames:
+            lookup = lookup[lookup.container_idx > -1]
+
+        Use case:
+        Given a Pupil timestamp, one can use bisect to find the corresponding
+        lookup entry index. From there, one can lookup the corresponding
+        container, load it if necessary, and calculate the target PTS
+        """
+        loaded_ts = self._loaded_ts_sorted()
+        filled_ts = self._fill_gaps(loaded_ts)
+        lookup = self._setup_lookup(filled_ts)
+        lookup.timestamp = filled_ts
+        lookup.container_idx = -1  #
+        for container_idx, vid in enumerate(self.videos):
+            mask = np.isin(lookup.timestamp, vid.timestamps)
+            lookup.container_frame_idx[mask] = np.arange(vid.timestamps.size)
+            try:
+                vid.check_container()  # check if container is corrupt
+                lookup.container_idx[mask] = container_idx
+            except av.AVError:
+                logger.warning(f"{vid.name} is corrupt")
+        self.lookup = lookup
+
+    def save_lookup(self):
+        np.save(self.lookup_loc, self.lookup)
+
+    def load_lookup(self):
+        self.lookup = np.load(self.lookup_loc).view(np.recarray)
+
+    def _loaded_ts_sorted(self) -> np.ndarray:
+        loaded_ts = [vid.timestamps for vid in self.videos]
+        all_ts = np.concatenate(loaded_ts)
```

Comment:
We don't have the sort all_ts here, right?

Reply:
We do since `_fill_gaps()` as well as the following code expect a sorted list.